device/platforms/stm32/GPIODevice.h

```diff
@@ -0,0 +1,100 @@
+/**
+ * HAL Implementation for GPIO Devices
+ *
+ * @author Aaron Chan
+ */
+
+#ifndef LAUNCH_CORE_GPIODEVICE_H
+#define LAUNCH_CORE_GPIODEVICE_H
+
+#include "stm32f4xx_hal_gpio.h"
+#include "stm32f4xx_hal_def.h"
+#include "stm32f446xx.h"
+
+
+#include "device/GPIODevice.h"
+#include "HAL_Handlers.h"
+#include "sched/sched.h"
+#include "sync/BlockingSemaphore.h"
+
+class HALGPIODevice : public GPIODevice, public CallbackDevice {
+public:
+
+    HALGPIODevice(const char *name, GPIO_TypeDef *halGPIO, uint16_t pin) : halGPIO(halGPIO),
+                                                                           pin(pin),
+                                                                           currentBlocked(-1),
+                                                                           taskLock(1),
+                                                                           GPIODevice(name) {};
+
+
+    RetType init() override {
+        return HALHandlers::register_gpio(halGPIO, this, REG_NUM);
+    }
+
+    RetType obtain() override {
+        return RET_SUCCESS;
+    }
+
+    RetType poll() override {
+        return RET_SUCCESS;
+    }
+
+    RetType release() override {
+        return RET_SUCCESS;
+    }
+
+    RetType set(uint32_t val) override {
+        RESUME();
+
+        RetType ret = CALL(taskLock.acquire());
+        RET_CHECK(ret);
+
+        taskLock = sched_dispatched;
+
+        HAL_GPIO_WritePin(halGPIO, this->pin, val);
+        BLOCK();
+
+        currentBlocked = -1;
+
+        ret = CALL(taskLock.release());
+        RET_CHECK(ret);
+
+        RESET();
+        return RET_SUCCESS;
+    }
+
+    RetType get(uint32_t *val) override {
+        RESUME();
+
+        RetType ret = CALL(taskLock.acquire());
+        RET_CHECK(ret);
+
+        taskLock = sched_dispatched;
+        *val = HAL_GPIO_ReadPin(halGPIO, pin);
+
+        BLOCK();
+
+        currentBlocked = -1;
+
+        ret = CALL(taskLock.release());
+        RET_CHECK(ret);
+
+        RESET();
+        return RET_SUCCESS;
+    }
+
+    void callback(int) override {
+        if (currentBlocked != -1) WAKE(currentBlocked);
+    }
+
+private:
+    static const int REG_NUM = 0;
+    tid_t currentBlocked;
+    alloc::Queue<tid_t, MAX_NUM_TASKS> taskQueue;
```

Comment:
what does this do?

Reply:
I think you copied this from stm32 SPI and I2C device, it can be removed on all of them. What this did is now implemented by BlockingSemaphore

Reply:
also may be good to take a look at BlockingSemphore to see why its used and why its abstracted out to its own class (lots of devices need this)

Reply:
I removed the queues from I2C, SPI, and UART controllers